### ISSUE TITLE
CommonsChunkPlugin in async mode doesn't select initial chunks caused a bug in expected behvaior

### ISSUE THREAD

**Issue by kazazor** [webpack/webpack#5109]
**Do you want to request a *feature* or report a *bug*?**
Bug

**What is the current behavior?**

The change that was made by [this PR](https://github.com/webpack/webpack/pull/4815) caused a change in expected behavior which IMO seems like a bug.

**If the current behavior is a bug, please provide the steps to reproduce.**

I have this configuration:
```js
entry: {
    application: [path.join(chunksPath, 'application_chunk')]
  },
....
plugins: {
  new webpack.optimize.CommonsChunkPlugin({
      async: "vendors",
      children: true,
      minChunks: (module, count) => {
        return module.resource &&
          webpackUtils.isVendor(module.resource) &&
          !webpackUtils.isPolyfill(module.resource) &&
          !webpackUtils.isSharedComponent(module.resource);
      }
    })
}
```

As you can see I was preparing an async vendors chunk if the module was in `node_modules` (`isVendor`), not a polufill (from `core-js`) and not a specific package we have that changes a lot and we consume it as a npm package.

In the initial chunk (`application`) there is very few little code and basically all it does is using `import()` to async require other chunks including out main chunk (we do that for running everything only after the polyfills were fetched).

This is the application chunk:
```js
import loadAllPolyfills from './polyfills_chunk';

const loadApplication = () => {
  import(/* webpackChunkName: "authentication_chunk" */ './authentication_chunk')
    .then(runAuth => runAuth.default())
    .then((user) => {
      import(/* webpackChunkName: "main_chunk" */ './main_chunk')
        .then(runMain => runMain.default(user))
        .catch((e) => {
          console.log('Error: ', e);
        });
    })
    .catch((e) => {
      console.log('Error: ', e);
    });
};

loadAllPolyfills()
  .then(loadApplication)
  .catch((e) => {
    console.log('Error: ', e);
    loadApplication();
  });
```

Before this change here this is how the main & vendors chunks looked:
<img width="1276" alt="screen shot 2017-05-30 at 12 08 02 am" src="https://cloud.githubusercontent.com/assets/1577655/26562105/1bba927a-44cc-11e7-8716-cde974a8b78b.png">

And after the change:
<img width="1222" alt="screen shot 2017-05-30 at 12 08 40 am" src="https://cloud.githubusercontent.com/assets/1577655/26562111/29aa9bb4-44cc-11e7-9d97-cce11ea4fd7d.png">

Basically the `vendors` chunk is empty now.
Like I said all that the initial chunk did was calling other async chunks. So there weren't any vendor modules there to begin with, so I find it hard to understand why this PR broke it for us..

**What is the expected behavior?**

It seems now that CommonChunkPlugin does not take any modules from async chunks what so ever.
And it used to take it.
If this was on purpose a way to do what was happening before should exists.

**Please mention other relevant information such as the browser version, Node.js version, webpack version and Operating System.**

OS: macOS
Node: 6.11.0 (happening also on 8.1.0)
Webpack: 2.5.0 and above (also on 3.0.0)

**Comment by aseem2625** [comment]
**Similar issue I faced:** https://github.com/webpack/webpack/issues/5101#issuecomment-309731764

**Comment by kazazor** [comment]
@aseem2625 - I tried adding what you added:
```js
    new webpack.optimize.CommonsChunkPlugin({
      names: ['vendors', 'manifest'],
      minChunks: Infinity
    }),
```

But now I'm getting this error:
```js
ERROR in CommonsChunkPlugin: While running in normal mode it's not allowed to use a non-entry chunk (vendors)
```

Any ideas?

**Comment by aseem2625** [comment]
Hey @kazazor I'm unware of your rest of the setup. But my point was that I didn't have to modify my existing system while upgrading to webpack v3 except that I have to update dependency of html webpack plugin because it was messing with commonChunks plugin which I'm using in similar fashion as yours.`

Now, let's see yours.
1) Why are you using `Infinity` there? The code block that you gave isn't same as mine. I'm using a function as value of `minChunks` (and note in the link I gave, I have added something for `manifest.js` as well)
2) Or alternatively you can try just
```
    new webpack.optimize.CommonsChunkPlugin({
      names: ['vendors', 'manifest']
    })
```

I think this error isn't related to webpack v3.
```
ERROR in CommonsChunkPlugin: While running in normal mode it's not allowed to use a non-entry chunk (vendors)
```

Do you mind giving link to a boilerplate of your setup? I can have a look at it and see where I can help.

Or have a look at this
https://stackoverflow.com/questions/37800009/webpack-how-to-use-commonschunkplugin-with-multiple-splitted-bundles

**Comment by iNikNik** [comment]
@kazazor 
just set name to your entry point ('application') like so:
```javasctipt
entry: {
    application: [path.join(chunksPath, 'application_chunk')]
  },
....
plugins: {
  new webpack.optimize.CommonsChunkPlugin({
      name: 'application',
      async: "vendors",
      children: true,
      minChunks: (module, count) => {
        return module.resource &&
          webpackUtils.isVendor(module.resource) &&
          !webpackUtils.isPolyfill(module.resource) &&
          !webpackUtils.isSharedComponent(module.resource);
      }
    })
}
```

I'm facing the same issue (without specifing entry point as name option - plugin does nothing).

## This is very unexpected behaivour becaus docs says:
```javascript
  name: string, // or
  names: string[],
  // The chunk name of the commons chunk. An existing chunk can be selected by passing a name of an existing chunk.
  // If an array of strings is passed this is equal to invoking the plugin multiple times for each chunk name.
  // If omitted and `options.async` or `options.children` is set all chunks are used,
  // otherwise `options.filename` is used as chunk name.```

**Comment by kazazor** [comment]
@iNikNik That's it!!

Thank you very much!
The docs are confusing here in this case..

**Comment by lijinke666** [comment]
good Thanks

**Comment by aseem2625** [comment]
@kazazor @iNikNik 
That's strange that changing entry point worked for him. Mine is like:

```
entry: {
	app: './src/client.js'
},

```

**Comment by kazazor** [comment]
I didn't change the entry point.. I just marked the entry point chunk as the chunk name for CommonChunkPlugin

It is strange due to the docs around it..
maybe @sokra can explain it since he was the one we performed the PR the changed the behavior..

**Comment by iNikNik** [comment]
@kazazor @sokra 2.5.0 release notes says:
> CommonsChunkPlugin in async mode doesn't extract from initial chunks

But docs says:

> If omitted and `options.async` or `options.children` is set all chunks are used

For me personally `async & children` should use really all cunks (including entry chunks) - because it's common case: many of us can read articles like this one https://medium.com/webpack/webpack-bits-getting-the-most-out-of-the-commonschunkplugin-ab389e5f318 ... but now it is very hard to make it works properly.
